src/lsp4clj/server.clj

```diff
@@ -31,44 +31,35 @@
 (defprotocol IBlockingDerefOrCancel
   (deref-or-cancel [this timeout-ms timeout-val]))
 
-(defrecord PendingRequest [p cancelled? id method started server]
+(defrecord PendingRequest [p id method started]
   clojure.lang.IDeref
   (deref [_] (deref p))
   clojure.lang.IBlockingDeref
   (deref [_ timeout-ms timeout-val]
     (deref p timeout-ms timeout-val))
   IBlockingDerefOrCancel
-  (deref-or-cancel [this timeout-ms timeout-val]
-    (let [result (deref this timeout-ms ::timeout)]
-      (if (= ::timeout result)
-        (do (future-cancel this)
+  (deref-or-cancel [_ timeout-ms timeout-val]
+    (let [result (deref p timeout-ms ::timeout)]
+      (if (identical? ::timeout result)
+        (do (p/cancel! p)
             timeout-val)
         result)))
   clojure.lang.IPending
-  (isRealized [_] (realized? p))
+  (isRealized [_] (p/done? p))
   java.util.concurrent.Future
-  (get [this]
-    (let [result (deref this)]
-      (if (= ::cancelled result)
-        (throw (java.util.concurrent.CancellationException.))
+  (get [_] (deref p))
+  (get [_ timeout unit]
+    (let [result (deref p (.toMillis unit timeout) ::timeout)]
+      (if (identical? ::timeout result)
+        (throw (java.util.concurrent.TimeoutException.))
         result)))
-  (get [this timeout unit]
-    (let [result (deref this (.toMillis unit timeout) ::timeout)]
-      (case result
-        ::cancelled (throw (java.util.concurrent.CancellationException.))
-        ::timeout (throw (java.util.concurrent.TimeoutException.))
-        result)))
-  (isCancelled [_] @cancelled?)
-  (isDone [this] (or (.isRealized this) (.isCancelled this)))
-  (cancel [this _interrupt?]
-    (if (.isDone this)
-      false
-      (if (compare-and-set! cancelled? false true)
-        (do
-          (protocols.endpoint/send-notification server "$/cancelRequest" {:id id})
-          (deliver p ::cancelled)
-          true)
-        false))))
+  (isCancelled [_] (p/cancelled? p))
+  (isDone [_] (p/done? p))
+  (cancel [_ _interrupt?]
+    (p/cancel! p)
+    (p/cancelled? p))
```

Comment:
I think `p/cancelled?` is unnecessary, as `p/cancel!` will already return the return value of the underlying `CompleteableFuture`, which itself implements `Future`. We can just do `(cancel [_ _] (p/cancel! p))`

Reply:
Unfortunately promesa, like similar libraries, is not very good at documenting return values of these operations.

Reply:
I don't think that's equivalent. `p/cancel!` returns the CompletableFuture object, but the Future interface says `.cancel` returns a boolean.

Reply:
Ah, you are right! I might have overlooked that `p/cancel` calls the implementation, ignores its return value, and returns the promise. Or maybe I looked at a different version of promesa...